### PR TITLE
Azure Idle Compute Instances

### PR DESCRIPTION
**Pull request by srpomeroy**
### Description

Report on any instances that have lower than specified average CPU utilization

### Issues Resolved

[List any existing issues this PR resolves]

### Contribution Check List

- [x] All tests pass.
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable
- [x] New functionality has been documented in CHANGELOG.MD

### CODE REVIEW

File: compliance/azure/azure_low_cpu_usage/Azure CPU Below Utilization.pt [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+  iterate $ds_azure_virtualmachines
+  request do
+    auth $azure_auth
+    pagination $azure_pagination
+    host "management.azure.com"
+    path join(["/subscriptions/", $param_azure_sub, "/resourceGroups/", val(iter_item,"rg"), "/providers/Microsoft.Compute/virtualMachines/",val(iter_item,"name"),"/providers/microsoft.insights/metrics"])
+    query "api-version","2016-09-01"
+    header "User-Agent", "RS Policies"
+  end 
+  result do
```
Comment (rshade) [inline compliance/azure/azure_low_cpu_usage/Azure CPU Below Utilization.pt]: Fix formatting

File: compliance/azure/azure_low_cpu_usage/Azure CPU Below Utilization.pt [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+         end
+  end
+end
+
+datasource "ds_combineddata" do
+  run_script $js_combineddata,$param_cpuaveragepercentage,$ds_azure_instance_performance
+end
+
+#This is to get the average cpu over time for the machine and return the dataset of machines with issues
+script "js_combineddata", type: "javascript" do
```
Comment (rshade) [inline compliance/azure/azure_low_cpu_usage/Azure CPU Below Utilization.pt]: Fix Formatting

File: compliance/azure/azure_low_cpu_usage/Azure CPU Below Utilization.pt [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+  var averagecpu 
+  for (var z = 0; z < averages.length; z++){
+    var averagenumber =averages[z]
+    averagecpu=0
+    flagtrip=0
+    if (!isNaN(averagenumber["average"]))
+      {
+    averagecpu = averagenumber["average"] + averagecpu
+    flagtrip=1
+  }
```
Comment (rshade) [inline compliance/azure/azure_low_cpu_usage/Azure CPU Below Utilization.pt]: Formatting

File: compliance/azure/azure_low_cpu_usage/Azure CPU Below Utilization.pt [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+policy "azure_resource_policy" do
+  validate $ds_combineddata do
+    summary_template "Azure Resources out of compliance"
+    detail_template <<-EOS
+# Azure Resources out of compliance
+| Name | CPU Average | Resource Group
+| ----------------------- | -----------------------  | -------------------- 
+{{ range data -}}
+{{.vmname}} | {{.averagecpu}} | {{.rg}}
```
Comment (rshade) [inline compliance/azure/azure_low_cpu_usage/Azure CPU Below Utilization.pt]: Initial and ending | need to be added

File: compliance/azure/azure_low_cpu_usage/Azure CPU Below Utilization.pt [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+          "content-type": "application/json",
+          "authorization": "Bearer " + $access_token
+        }
+      )
+      $all_responses << $response
+    end
+  end
+end
+
+escalation "delete_resources" do
```
Comment (rshade) [inline compliance/azure/azure_low_cpu_usage/Azure CPU Below Utilization.pt]: Group like items together please

File: compliance/azure/azure_low_cpu_usage/Azure CPU Below Utilization.pt [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        https: true,
+        query_strings: {
+          "api-version": "2018-06-01"
+        },
+        headers: {
+          "cache-control": "no-cache",
+          "content-type": "application/json",
+          "authorization": "Bearer " + $access_token
+        }
+      )
```
Comment (rshade) [inline compliance/azure/azure_low_cpu_usage/Azure CPU Below Utilization.pt]: log the response, and treat error to log definition.

File: compliance/azure/azure_low_cpu_usage/README.md [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+### Functional Details
+ 
+The policy leverages the Azure API to check all instances and then grab the utilization over the past hour, finally recomendending the low ones for decomissioning
+ 
+#### Input Parameters
+ 
+- *Email addresses of the recipients you wish to notify* - A list of email addresses to notify
+- *CPU Threshold* - Percentage of CPU utilization
+- *Azure Subscription ID* - Your Azure Subscription ID.  You can find it by following this guide: https://docs.bitnami.com/azure/faq/administration/find-subscription-id/
```
Comment (rshade) [inline compliance/azure/azure_low_cpu_usage/README.md]: Check other azure policies for this parameter and match them.

File: compliance/azure/azure_low_cpu_usage/Azure CPU Below Utilization.pt (hunk 4 above)
Comment (LegacyAcres) [reply compliance/azure/azure_low_cpu_usage/Azure CPU Below Utilization.pt]: Intitial and ending what?

File: compliance/azure/azure_low_cpu_usage/Azure CPU Below Utilization.pt (hunk 5 above)
Comment (LegacyAcres) [reply compliance/azure/azure_low_cpu_usage/Azure CPU Below Utilization.pt]: Done

File: compliance/azure/azure_low_cpu_usage/README.md (hunk 7 above)
Comment (LegacyAcres) [reply compliance/azure/azure_low_cpu_usage/README.md]: It seems to match

File: compliance/azure/azure_low_cpu_usage/Azure CPU Below Utilization.pt (hunk 6 above)
Comment (LegacyAcres) [reply compliance/azure/azure_low_cpu_usage/Azure CPU Below Utilization.pt]: Done

File: compliance/azure/azure_low_cpu_usage/Azure CPU Below Utilization.pt [hunk 8]
```diff
@@ -0,0 +1,256 @@
+name "Azure instances with low cpu average"
```
Comment (dfrankel33) [inline compliance/azure/azure_low_cpu_usage/Azure CPU Below Utilization.pt]: rename to "Azure Idle Compute Instances"

File: compliance/azure/azure_low_cpu_usage/README.md [hunk 9]
```diff
@@ -0,0 +1,27 @@
+## Azure Below CPU Utilization
```
Comment (dfrankel33) [inline compliance/azure/azure_low_cpu_usage/README.md]: rename to "Azure Idle Compute Instances"

File: compliance/azure/azure_low_cpu_usage/Azure CPU Below Utilization.pt (hunk 8 above)
Comment (LegacyAcres) [reply compliance/azure/azure_low_cpu_usage/Azure CPU Below Utilization.pt]: Ill rename it Azure Idle Compute Instances lol

File: compliance/azure/azure_idle_compute_instances/README.md [hunk 10]
```diff
@@ -0,0 +1,27 @@
+## Azure Idle Compute Instances
+ 
+### What it does
+This policy checks all the instances in the Azure Subscription for the average CPU usage over the the last hour.  If the usage is less than the user provided CPU percentage threshold then the virtual machines are recomended for removal, and the user is emailed.
```
Comment (cdwilhelm) [inline compliance/azure/azure_idle_compute_instances/README.md]: use the same terminology used in the template description and approval.  review the terms used for the actions and pick one,  the policy and readme include removal, termination and decommission.  use one for consistency and avoiding confusing.

File: compliance/azure/azure_idle_compute_instances/README.md [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+#### Input Parameters
+ 
+- *Email addresses of the recipients you wish to notify* - A list of email addresses to notify
+- *CPU Threshold* - Percentage of CPU utilization
+- *Azure Subscription ID* - Your Azure Subscription ID.  You can find it by following this guide: https://docs.bitnami.com/azure/faq/administration/find-subscription-id/
+- *Azure Tenant ID* - Your Azure tenant ID.  You can find it by following this guide: https://docs.microsoft.com/en-us/onedrive/find-your-office-365-tenant-id
+ 
+### Required RightScale Roles
+ 
+None
```
Comment (cdwilhelm) [inline compliance/azure/azure_idle_compute_instances/README.md]: need policy manger and approval.  See other policies for examples.

File: compliance/azure/azure_idle_compute_instances/Azure Idle Compute Instances.pt [hunk 12]
```diff
@@ -0,0 +1,275 @@
+name "Azure Idle Compute Instances"
+rs_pt_ver 20180301
+type "policy"
+short_description "Report on any instances that have lower than specified average CPU utilization"
```
Comment (cdwilhelm) [inline compliance/azure/azure_idle_compute_instances/Azure Idle Compute Instances.pt]: should include details they will be terminated after approval.

File: compliance/azure/azure_idle_compute_instances/README.md [hunk 13]
```diff
@@ -0,0 +1,27 @@
+## Azure Idle Compute Instances
+ 
+### What it does
+This policy checks all the instances in the Azure Subscription for the average CPU usage over the the last hour.  If the usage is less than the user provided CPU percentage threshold then the virtual machines are recomended for removal, and the user is emailed.
+
+### Functional Details
+ 
+The policy leverages the Azure API to check all instances and then grab the utilization over the past hour, finally recomendending the low ones for decomissioning
```
Comment (cdwilhelm) [inline compliance/azure/azure_idle_compute_instances/README.md]: include the approval details.

File: compliance/azure/azure_idle_compute_instances/README.md (hunk 10 above)
Comment (cdwilhelm) [reply compliance/azure/azure_idle_compute_instances/README.md]: there is an extra THE in the first paragraph

Comment (cdwilhelm) [reply compliance/azure/azure_idle_compute_instances/README.md]: Is the metric really on the last hour.  That seems very aggressive.  Can we make it 30 days so it's consistent with the utilization policy?

File: compliance/azure/azure_idle_compute_instances/README.md (hunk 11 above)
Comment (srpomeroy) [reply compliance/azure/azure_idle_compute_instances/README.md]: @cdwilhelm We aren't adding these anymore. We do however need to add credential_viewer.

File: compliance/azure/azure_idle_compute_instances/README.md [hunk 14]
```diff
@@ -0,0 +1,32 @@
+# Azure Idle Compute Instances
+
+## What it does
+
+This policy checks all the instances in the Azure Subscription for the average CPU usage over the last hour.  If the usage is less than the user provided CPU percentage threshold then the virtual machines are recommended for deletion, and the user is emailed.
```
Comment (cdwilhelm) [inline compliance/azure/azure_idle_compute_instances/README.md]: looks like the api call is the last 30day of metrics, not last hour

File: compliance/azure/azure_idle_compute_instances/README.md [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+
+## Functional Details
+
+The policy leverages the Azure API to check all instances and then checks the instance average CPU utilization over the past hour, finally recommending the low ones for deletion after approval.
```
Comment (cdwilhelm) [inline compliance/azure/azure_idle_compute_instances/README.md]: https://cloud.google.com/sql/docs/

File: compliance/azure/azure_idle_compute_instances/azure_idle_compute_instances.pt [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+###############################################################################
+# Policy
+###############################################################################
+
+policy "azure_resource_policy" do
+  validate $ds_combineddata do
+    summary_template "{{ rs_project_name }} (Account ID: {{ rs_project_id }}): Azure idle compute instances found"
+    detail_template <<-EOS
+# Azure Resources out of compliance
+| Location | Name | CPU Average | Resource Group | Tags |
```
Comment (cdwilhelm) [inline compliance/azure/azure_idle_compute_instances/azure_idle_compute_instances.pt]: shouldn't we include the region and instance id like other polices

File: compliance/azure/azure_idle_compute_instances/azure_idle_compute_instances.pt [hunk 17]
```diff
@@ -0,0 +1,275 @@
...
+
+###############################################################################
+# Policy
+###############################################################################
+
+policy "azure_resource_policy" do
+  validate $ds_combineddata do
+    summary_template "{{ rs_project_name }} (Account ID: {{ rs_project_id }}): Azure idle compute instances found"
+    detail_template <<-EOS
+# Azure Resources out of compliance
```
Comment (cdwilhelm) [inline compliance/azure/azure_idle_compute_instances/azure_idle_compute_instances.pt]: How about using **Azure idle compute instances.**

File: cost/azure/idle_compute_instances/azure_idle_compute_instances.pt [hunk 18]
```diff
@@ -0,0 +1,277 @@
...
+  end
+  call sys_log(join([$syslog_subject, "Responses"]),to_s($all_responses))
+end
+
+
+
+define get_access_token($param_azure_tenant_id) return $access_token do
+
+  $client_id = cred("AZURE_APPLICATION_ID")
+  $client_secret = cred("AZURE_APPLICATION_KEY")
```
Comment (cdwilhelm) [inline cost/azure/idle_compute_instances/azure_idle_compute_instances.pt]: need to url_encode the secret.
https://github.com/flexera/policy_templates/blob/master/cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt#L511

File: cost/azure/idle_compute_instances/azure_idle_compute_instances.pt [hunk 19]
```diff
@@ -0,0 +1,324 @@
+name "Azure Idle Compute Instances"
+rs_pt_ver 20180301
+type "policy"
+short_description "Report on any instances that have lower than specified average CPU utilization and delete them with approval"
```
Comment (cdwilhelm) [inline cost/azure/idle_compute_instances/azure_idle_compute_instances.pt]: change short_description to: Checks for Azure Compute Instances that are idle for the last 30 days and terminates them after approval.

**Pull request merged**